fdbserver/Backup.actor.cpp

```diff
@@ -0,0 +1,262 @@
+				self->versions.erase(self->versions.begin());
+			}
+		}
+
+		// TODO: upload messages
+
+		state Future<Void> savedProgress = Never();
+		if (popVersion > self->savedVersion) {
+			savedProgress = saveProgress(self, popVersion);
+		}
+		// TODO: knobify the delay of 20s. This delay is sensitive, as it is the
+		// lag TLog might have.
+		state Future<Void> uploadDelay = delay(20);
+
+		loop choose {
+			when(wait(uploadDelay)) {
+				// Note the progress may not be saved yet. Cancel it?
+				break;
+			}
+			when(wait(savedProgress)) {
+				TraceEvent("BackupWorkerSavedProgress", self->myId).detail("Version", popVersion);
+				self->savedVersion = std::max(popVersion, self->savedVersion);
+				savedProgress = Never();  // Still wait until uploadDelay expires
+			}
+		}
+	}
+}
+
+// Pulls data from TLog servers using LogRouter tag.
+ACTOR Future<Void> pullAsyncData(BackupData* self) {
+	state Future<Void> logSystemChange = Void();
+	state Reference<ILogSystem::IPeekCursor> r;
+	state Version tagAt = self->startVersion;
+	state Version lastVersion = 0;
+
+	loop {
+		loop choose {
+			when (wait(r ? r->getMore(TaskPriority::TLogCommit) : Never())) {
+				break;
+			}
+			when (wait(logSystemChange)) {
+				if (self->logSystem.get()) {
+					r = self->logSystem.get()->peekLogRouter(self->myId, tagAt, self->tag);
+				} else {
+					r = Reference<ILogSystem::IPeekCursor>();
+				}
+				logSystemChange = self->logSystem.onChange();
+			}
+		}
+		self->minKnownCommittedVersion = std::max(self->minKnownCommittedVersion, r->getMinKnownCommittedVersion());
+
+		// TODO: avoid peeking uncommitted messages
+		// Should we wait until knownCommittedVersion == startVersion - 1 ? In this way, we know previous
+		// epoch has finished and then starting for this epoch.
+		while (r->hasMessage()) {
+			lastVersion = r->version().version;
+			self->messages.emplace_back(r->getMessage(), std::vector<Tag>());
+			self->versions.push_back(lastVersion);
+			r->nextMessage();
+		}
+
+		tagAt = std::max(r->version().version, lastVersion);
+		self->lastSeenVersion = std::max(tagAt, self->lastSeenVersion);
+		TraceEvent("BackupWorkerGot", self->myId).suppressFor(1.0).detail("V", tagAt);
+		if (tagAt > self->endVersion) {
+			TraceEvent("BackupWorkerFinishPull", self->myId)
+			    .detail("VersionGot", tagAt)
+			    .detail("EndVersion", self->endVersion);
+			return Void();
+		}
+		wait(yield());
+	}
+}
+
+// We need to pop data faster than the upload interval to avoid accumulate
+// mutations in TLogs.
+ACTOR Future<Void> popData(BackupData* self) {
+	loop {
+		if (self->logSystem.get()) {
+			const Tag popTag = self->logSystem.get()->getPseudoPopTag(self->tag, ProcessClass::BackupClass);
+			self->logSystem.get()->pop(self->savedVersion, popTag);
+		}
+		wait(delay(2.0));  // TODO: knobify this delay
+	}
+}
+
+ACTOR Future<Void> checkRemoved(Reference<AsyncVar<ServerDBInfo>> db, LogEpoch recoveryCount,
+                                BackupData* self) {
```

Comment:
change to include the backup interfaces in dbInfo, so we can check when they are removed to determine if we should die

Reply:
I am going to use a different logic. I.e., the backup worker, when done, notifies master to remove itself from log system. After master acknowledges, the backup worker then exits.

So there is no change here.